File: src/24-swapPairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
  if (!head || !head.next) {
    return head;
  }

  var node1 = head;
  var node2 = node1.next;
  var node3 = node2.next;

  node2.next = node1;
  node1.next = swapPairs(node3);

  return node2;
};

File: src/24-swapPairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
  public ListNode swapPairs(ListNode head) {
    if (head == null) {
      return head;
    }
    
    ListNode swappedListHead = new ListNode(-1, head);
    ListNode currentNode = swappedListHead.next;
    ListNode previousNode = swappedListHead;
  
    while (currentNode != null && currentNode.next != null) {
      ListNode nextNode = currentNode.next;
      previousNode.next = nextNode;
      currentNode.next = nextNode.next;
      nextNode.next = currentNode;
      previousNode = currentNode;
      currentNode = currentNode.next;
    }
  
    return swappedListHead.next;
  }
}
